pkg/registry/core/pod/storage/eviction.go

```diff
@@ -232,18 +246,48 @@ func (r *EvictionREST) Create(ctx context.Context, name string, obj runtime.Obje
 		return rtStatus, nil
 	}
 
-	// At this point there was either no PDB or we succeeded in decrementing
+	// At this point there was either no PDB or we succeeded in decrementing or
+	// the pod was unready and we have enough healthy replicas
+
+	deletionOptions := originalDeleteOptions.DeepCopy()
+
+	// Set deletionOptions.Preconditions.ResourceVersion to ensure
+	// the pod hasn't been considered ready since we calculated
+	if updateDeletionOptions {
+		setPreconditionsResourceVersion(deletionOptions, &pod.ResourceVersion)
+	}
 
 	// Try the delete
-	_, _, err = r.store.Delete(ctx, eviction.Name, rest.ValidateAllObjectFunc, originalDeleteOptions.DeepCopy())
+	_, _, err = r.store.Delete(ctx, eviction.Name, rest.ValidateAllObjectFunc, deletionOptions.DeepCopy())
 	if err != nil {
+		if errors.IsConflict(err) && updateDeletionOptions &&
+			(originalDeleteOptions.Preconditions == nil || originalDeleteOptions.Preconditions.ResourceVersion == nil) {
+			// If we encounter a resource conflict error, we updated the deletion options to include them,
+			// and the original deletion options did not specify ResourceVersion, we send back
+			// TooManyRequests so clients will retry.
+			return nil, createTooManyRequestsError(pdbName)
```

Comment:
Can you explain from history *why* we were returning TooManyRequests in this code base?  Was it a quick workaround to trigger retry?  Or was there a longer discussion about the appropriate response code.  I don't have confidence from the comments here that it's the right error code, and I'd want to be confident before we spread it to a second spot (and update the comments such that it's explicitly clear that no other status code is appropriate, if so)

Reply:
Here's the original implementing PR, this comment makes me believe that we chose an error number out of a hat: https://github.com/kubernetes/kubernetes/pull/31033#discussion_r75573684

Current state of the world as far as kubectl/drain is that while trying to evict a pod, if I get a 429, that indicates the PDB did not allow eviction at this time, retry.

We are sending 429 here to emulate a blocking PDB (eg, PDB with 0 disruptions available) as we do above.  This is because if we fail the ResourceVersion constraint here, we are no longer in position to evaluate the PDBs.

I think signaling the client to optionally retry is the most sensible option here, and fits the current behavior: If I had 0 disruptions left (because the pod I'm trying to evict is already disrupted) we send back a 429.

Reply:
Thinking about the status codes we've assigned in the past (as the rough originator of most of the error stanzas I probably should create on separately for this if we will use it in the future), there's a couple that are close but not equivalent:

If you specify `generateName` and we keep getting collisions - the server generates a random name and then attempts to create the object in etcd, which could fail because someone else could choose that name - after 10 we return ServerTimeout

```
// NewServerTimeout returns an error indicating the requested action could not be completed due to a
// transient error, and the client should try again.
```

Patch retries forever, as does a guaranteed update.

Reply:
I guess 429 is fine for now, I can't really think of a better one and I'd rather not couple this change to to a more dramatic client behavior change.